File: part1/CPP/main.cpp
#include<iostream>
#include <chrono>
#include<cstdlib>	
#include "arraylist.hpp"
#include "analyzer.hpp"


using namespace std::chrono;


int main(int argc, char ** argv){
  int i;
  // Tableau dynamique.
  ArrayList<int> a;
  // Analyse du temps pris par les opérations.
  Analyzer time_analysis;
  // Analyse du nombre de copies faites par les opérations.
  Analyzer copy_analysis;
  // Analyse de l'espace mémoire inutilisé.
  Analyzer memory_analysis;
  high_resolution_clock::time_point before, after;
  // Booléen permettant de savoir si une allocation a été effectuée.
  bool memory_allocation;

  for(i = 0; i < 1000000 ; i++){
    // Ajout d'un élément et mesure du temps pris par l'opération.
    before = high_resolution_clock::now();
    memory_allocation = a.append(i);
    after = high_resolution_clock::now();

    // Enregistrement du temps pris par l'opération
    auto duration = std::chrono::duration_cast<std::chrono::nanoseconds>(after - before);    
    time_analysis.append(static_cast<double>(duration.count()));
    // Enregistrement du nombre de copies efféctuées par l'opération.
    // S'il y a eu réallocation de mémoire, il a fallu recopier tout le tableau.
    copy_analysis.append( (memory_allocation)? i:1 );
    // Enregistrement de l'espace mémoire non-utilisé.
    memory_analysis.append( a.capacity() - a.size() );
  }

  // Affichage de quelques statistiques sur l'expérience.
  std::cerr<<"Total cost :"<<time_analysis.get_total_cost()<<std::endl;
  std::cerr<<"Average cost :"<<time_analysis.get_average_cost()<<std::endl;
  std::cerr<<"Variance :"<<time_analysis.get_variance()<<std::endl;
  std::cerr<<"Standard deviation :"<<time_analysis.get_standard_deviation()<<std::endl;    

  // Sauvegarde les données de l'expérience.
  time_analysis.save_values("../plots/dynamic_array_time_cpp.plot");
  copy_analysis.save_values("../plots/dynamic_array_copy_cpp.plot");
  memory_analysis.save_values("../plots/dynamic_array_memory_cpp.plot");
  
  return 0;
}

File: part1/Java/Main.java
import java.util.Random;

public class Main {

    public static void main(String[] args) {
        int i, time_ind = 0, pop_ind = 0;
        // Tableau dynamique.
        ArrayListProxy<Integer> a = new ArrayListProxy<Integer>();
        // Analyse du temps pris par les opérations.
        Analyzer time_analysis = new Analyzer();
        // Analyse du nombre de copies faites par les opérations.
        Analyzer copy_analysis = new Analyzer();
	// Analyse de l'espace mémoire inutilisé.
	Analyzer memory_analysis = new Analyzer();
        long before, after;
        // Booléen permettant de savoir si une allocation a été effectuée.
        boolean memory_allocation;


        for(i = 0; i < 1000000 ; i++){
            // Ajout d'un élément et mesure du temps pris par l'opération.
            before = System.nanoTime();
            memory_allocation = a.append(i);
            after = System.nanoTime();

            // Enregistrement du temps pris par l'opération
            time_analysis.append(after - before);
            // Enregistrement du nombre de copies efféctuées par l'opération.
            // S'il y a eu réallocation de mémoire, il a fallu recopier tout le tableau.
            copy_analysis.append( (memory_allocation == true)? i: 1);
	    // Enregistrement de l'espace mémoire non-utilisé.
	    memory_analysis.append( a.capacity() - a.size() );
        }

        // Affichage de quelques statistiques sur l'expérience.
        System.err.println("Total cost : "+time_analysis.get_total_cost());
        System.err.println("Average cost : "+time_analysis.get_average_cost());
        System.err.println("Variance :"+time_analysis.get_variance());
        System.err.println("Standard deviation :"+time_analysis.get_standard_deviation());

        // Sauvegarde les données de l'expérience: temps et nombre de copies effectuées par opération.
        time_analysis.save_values("../plots/dynamic_array_time_java.plot");
        copy_analysis.save_values("../plots/dynamic_array_copy_java.plot");
	memory_analysis.save_values("../plots/dynamic_array_memory_java.plot");
    }
}
